src/Core/HTTPApplication.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+namespace SilverStripe\Core;
+
+use SilverStripe\Control\Director;
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\HTTPResponse;
+
+/**
+ * Invokes the HTTP application within an ErrorControlChain
+ */
+class HTTPApplication implements Application
+{
+    /**
+     * @var callable[]
+     */
+    protected $middlewares = [];
+
+    /**
+     * @return callable[]
+     */
+    public function getMiddlewares()
+    {
+        return $this->middlewares;
+    }
+
+    /**
+     * @param callable[] $middlewares
+     * @return $this
+     */
+    public function setMiddlewares($middlewares)
+    {
+        $this->middlewares = $middlewares;
+        return $this;
+    }
+
+    /**
+     * @param callable $middleware
+     * @return $this
+     */
+    public function addMiddleware($middleware)
+    {
+        $this->middlewares[] = $middleware;
+        return $this;
+    }
+
+    /**
+     * Call middleware
+     *
+     * @param callable $last Last config to call
+     * @return HTTPResponse
+     */
+    protected function callMiddleware($last)
```

Comment:
We've added the request to the signature, following the PSR-15 middleware API. It's not a direct implementation, but follows the same method arguments / ordering. Once we implement PSR-7 it should be a simple shift to also support PSR-15.